### ISSUE TITLE
Enable outside air infiltration in Spawn for zones not modeled with EnergyPlus.ThermalZone

### ISSUE THREAD

**Issue by mwetter** [lbl-srg/modelica-buildings#2492]
This issue is to make changes to how infiltration is handled in Spawn for zones that are not connected.

Currently, all infiltration is removed by Spawn. While this is correct for zones that are simulated in Modelica (e.g., "connected zones"), it can lead to unreasonable temperatures in cases where a zone is unconditioned and hence a user may not connect it to an instance of `Buildings.ThermalZones.EnergyPlus.ThermalZone`. A use case is an unconditioned attic, such as used in `Buildings.ThermalZones.EnergyPlus.Examples.SingleFamilyHouse.RadiantHeatingCooling`.

- [x] Once infiltration is modeled in unconnected zones, then `Buildings.ThermalZones.EnergyPlus.UsersGuide.Conventions` need to be updated with text such as "If a zone is not modeled in Modelica, then if the idf file specifies infiltration for this zone, EnergyPlus will model it. (This allows for example to model an unconditioned attic with infiltration in EnergyPlus and model the rest of the building in Modelica.)"

It is not yet decided whether this change should be done prior to Release 8.0.0

**Comment by mwetter** [comment]
@kbenne : 
The new binaries from https://github.com/NREL/Spawn/issues/31#issuecomment-842796700 do not seem to model infiltration in unconnected zones. I added the test `Buildings.ThermalZones.EnergyPlus.Validation.ThermalZone.Infiltration` on branch `issue2492_spawnInfiltration`, commit https://github.com/lbl-srg/modelica-buildings/commit/e1b7403375010767a4106c56fb1a556535bed376 to compare a building with an attic with high and low infiltration.

Comparing the surface temperatures of the floor of the attic between a case with normal and high infiltration in the attic shows that the surface temperatures are identical. I expect different temperatures. The output is

![image](https://user-images.githubusercontent.com/1329080/118657459-5b56d600-b7a0-11eb-864a-911f01063c7a.png)

Do you get different results in your CI tests?

**Comment by kbenne** [comment]
Can you take a look at the idf file in the simulation directory and see if `ZoneInfiltration:DesignFlowRate` objects are in place for the unconnected zones? I don't have any automated tests beyond confirming that the infiltration input is left in place. I do not have quantitative confirmation.

**Comment by mwetter** [comment]
Here is the line with the object in place: https://github.com/lbl-srg/modelica-buildings/blob/e1b7403375010767a4106c56fb1a556535bed376/Buildings/Resources/Data/ThermalZones/EnergyPlus/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance_aboveSoilHighInfiltration.idf#L1186

Can you please add a test in the E+ CI that tests for changes in result. We need to catch such issues early on and guard against future revisions accidentally breaking it.

**Comment by kbenne** [comment]
I meant the version of the idf after it has been through spawn's processing. I expect the infiltration related to the living space (connected I think) to be removed, but the attic and garage infiltration to be left in place. I suspect this is so and the root cause is more complicated, but a good sanity check.

**Comment by mwetter** [comment]
Yes, `ZoneInfiltration:DesignFlowRate` is in place for the attic and garage, but not the living room.

**Comment by kbenne** [comment]
> Can you please add a test in the E+ CI that tests for changes in result. We need to catch such issues early on and guard against future revisions accidentally breaking it.

I certainly want that, but regression testing isn't something I can implement quickly or casually. It is going to take some effort. EnergyPlus has a very good system in place, and my inclination is to borrow as much of that as possible. In the mean time I am going to look at adding a test that requests an infiltration output variable and checks for a non zero result. However, that probably won't catch what I believe is the root cause which is that up until this point the zone heat balance sums have deliberately excluded terms that relate to air movement.

**Comment by kbenne** [comment]
There is a duplicate issue reported on the NREL/spawn repository. https://github.com/NREL/Spawn/issues/31 

I referenced a potential fix for this issue https://github.com/NREL/Spawn/commit/5a19a3aaeac0b55a6f063158fca80a680af6bb07, however I'm a little blind to the full impacts until there is some kind of regression testing setup. We should proceed with caution.

**Comment by kbenne** [comment]
This is the build you'll want https://gitlab.com/kylebenne/spawn/-/pipelines/305438660

**Comment by mwetter** [comment]
@kbenne : I tried the above build but the results are identical for the case shown in https://github.com/lbl-srg/modelica-buildings/issues/2492#issuecomment-843190866

This is the version I tested:
```
--- a/Buildings/Resources/src/ThermalZones/EnergyPlus/install.py
+++ b/Buildings/Resources/src/ThermalZones/EnergyPlus/install.py
@@ -155,7 +155,7 @@ if __name__ == "__main__":
     # Commit, see https://gitlab.com/kylebenne/spawn/-/pipelines?scope=all&page=1
     # Also available is latest/Spawn-latest-{Linux,win64,Darwin}
     # The setup below lead to a specific commit being pulled.
-    commit = "09f8cff897fac3aa6054e69f4c6d4b9bc923d2db"
+    commit = "94eecd9b62dcd457f8526d3eaa87c213d3f44420"
     name_version = f"Spawn-0.1.0-{commit[0:10]}"
```
The Buildings library branch is at commit c06ba2139b8e4496f6b4a8329590e714060eec6f, and before running the example, `Resources/src/ThermalZones/EnergyPlus/install.py` needs to be run as I did not commit the large binaries for this iteration.

Do you get different results if you make the change below in an idf file?
```
Buildings$ diff Resources/Data/ThermalZones/EnergyPlus/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance_aboveSoil.idf Resources/Data/ThermalZones/EnergyPlus/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance_aboveSoilHighInfiltration.idf
1186c1186
<     0.05,                    !- Design Flow Rate {m3/s}
---
>     1.,                      !- Design Flow Rate {m3/s}  -- high infiltration rate for model testing

```

**Comment by kbenne** [comment]
OK. thanks for testing. I'll do some analysis and try to devise a test to confirm that infiltration happens.

**Comment by kbenne** [comment]
Alright. here is another commit https://github.com/NREL/Spawn/commit/d60b4e277b8d3ed6e637949eee4e8df39eae69da that adds one more piece of the puzzle. I added a necessary call to CalcAirFlowSimple which populates the infiltration terms that then feed into the zone heat sums. 

Importantly, in this commit there is a new test that conclusively shows that infiltration is now impacting the simulation in the direction that I would expect. You bring in cold winter air and the zone gets colder, relative an identical zone without the infiltration. 

Here again, the biggest concern to watch out for are regressions.

**Comment by kbenne** [comment]
Build will be this one https://gitlab.com/kylebenne/spawn/-/pipelines/306842615

**Comment by mwetter** [comment]
@kbenne :
The build https://gitlab.com/kylebenne/spawn/-/pipelines/306842615 does not produce the expected results.
Simulating `Buildings.ThermalZones.EnergyPlus.Validation.ThermalZone.TwoIdenticalZones` shows that in connected zones, infiltration is still modeled, which is counter to https://github.com/lbl-srg/modelica-buildings/blob/079b2a60b3f0d12df5e8762406638bbfbeec07d7/Buildings/ThermalZones/EnergyPlus/UsersGuide.mo#L182

Moreover, if I add an output variable (in 079b2a60b3) that reports infiltration, e.g.,
```modelica
  Buildings.ThermalZones.EnergyPlus.OutputVariable inf1(
    name="Zone Infiltration Current Density Volume Flow Rate",
    key="Thermal Zone 1")
    annotation (Placement(transformation(extent={{40,-40},{60,-20}})));
```
the reported output is zero. But if I change the idf file as in
```
$ git diff Resources/Data/ThermalZones/EnergyPlus/Validation/TwoIdenticalZones/TwoIdenticalZones.idf
diff --git a/Buildings/Resources/Data/ThermalZones/EnergyPlus/Validation/TwoIdenticalZones/TwoIdenticalZones.idf b/Buildings/Resources/Data/ThermalZones/EnergyPlus/Validation/TwoIdenticalZones/TwoIdenticalZones.idf
index 8d7c27acda..c954a6f4c1 100644
--- a/Buildings/Resources/Data/ThermalZones/EnergyPlus/Validation/TwoIdenticalZones/TwoIdenticalZones.idf
+++ b/Buildings/Resources/Data/ThermalZones/EnergyPlus/Validation/TwoIdenticalZones/TwoIdenticalZones.idf
@@ -1090,7 +1090,7 @@
     ,                        !- Design Flow Rate {m3/s}
     ,                        !- Flow per Zone Floor Area {m3/s-m2}
     ,                        !- Flow per Exterior Surface Area {m3/s-m2}
-    0.3,                     !- Air Changes per Hour {1/hr}
+    0.,                     !- Air Changes per Hour {1/hr}
     1,                       !- Constant Term Coefficient
     0,                       !- Temperature Term Coefficient
     0,                       !- Velocity Term Coefficient
@@ -1104,7 +1104,7 @@
     ,                        !- Design Flow Rate {m3/s}
     ,                        !- Flow per Zone Floor Area {m3/s-m2}
     ,                        !- Flow per Exterior Surface Area {m3/s-m2}
-    0.3,                     !- Air Changes per Hour {1/hr}
+    0.,                     !- Air Changes per Hour {1/hr}
     1,                       !- Constant Term Coefficient
     0,                       !- Temperature Term Coefficient
     0,                       !- Velocity Term Coefficient
@@ -1160,4 +1160,3 @@
 
   Output:Diagnostics,
     DisplayAllWarnings;      !- Key 1
-
```
to set the infiltration to zero, I get the previous results. So the diff is due to the infiltration being added to the energy balance.

In summary, EnergyPlus reports that there is no infiltration, but it still adds the infiltration to the energy balance of the zone that is modeled in Modelica.

**Comment by kbenne** [comment]
I'm working from the following test https://github.com/NREL/Spawn/blob/issue/31-infiltration/test/test_infiltration.cpp. I will build on this test to capture some of what you have described here.

**Comment by kbenne** [comment]
The previous commit gets at part of what you've (@mwetter) described. Specifically the output variables related to infiltration should now be updated and a test has been added for that. (although you might want to read through the comments in the test)

I'll take a look at connected zones now. The infiltration input should be completely removed from the idf for connected zones. I confirmed that with manual inspect but do not have a formal test yet. I'm going to add some kind of formal test and see what I uncover.

 I want to clarify, that "connected" means that the zone is identified in the spawn input file....

```
 "model": {{
        "zones": [
           {{ "name": "Thermal Zone 1" }} 
        ]
      }}
```

which from the Modelica side means there is a corresponding Zone model instance.

**Comment by kbenne** [comment]
Update. The connected zone problem is because the two identical zones idf uses zone lists. sigh...

**Comment by kbenne** [comment]
This is an annoying problem to say the least.

```

ZoneInfiltration:DesignFlowRate,
  189.1-2009 - Office - WholeBuilding - Md Office - CZ4-8 Infiltration,
  Space Type 1,
  Infiltration Sched,
  AirChanges/Hour,
  ,
  ,
  ,
  0.3,
  1.0,
  0.0,
  0.0,
  0.0;

ZoneInfiltration:DesignFlowRate,
  189.1-2009 - Office - WholeBuilding - Md Office - CZ4-8 Infiltration 1,
  Space Type 2,
  Infiltration Sched,
  AirChanges/Hour,
  ,
  ,
  ,
  0.3,
  1.0,
  0.0,
  0.0,
  0.0;

ZoneList,
  Space Type 1,
  Thermal Zone 1;

ZoneList,
  Space Type 2,
  Thermal Zone 2;

```

There is one zone per zone list in this case, so it would be easily addressed, but I guess I'm going to need to account for it not being the case. potentially undoing the use of zone list by duplicating the infiltration input object.

**Comment by kbenne** [comment]
and this is an openstudio generated model no less, meaning the issue could be prevalent.

**Comment by mwetter** [comment]
@kbenne : Correct, a zone is called "connected" if and only if it is identified in the spawn input file, such as in
```
 "model": {
        "zones": [
           { "name": "Thermal Zone 1" } 
        ]
      }
```

**Comment by kbenne** [comment]
@mwetter here is the build https://gitlab.com/kylebenne/spawn/-/pipelines/308722155
